<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Marguerite Butler">
<meta name="dcterms.date" content="2025-04-23">
<meta name="description" content="Working with trees with tidytree and plotting with ggtree">
<title>The ggtree-verse â€“ ZOOL710 Data Science in R for Biologists 2025</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-2c128a76c5dbb243be50ea3089ffe19d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="../../styles.css">
</head>
<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">ZOOL710 Data Science in R for Biologists 2025</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-general-information" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">General Information</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-general-information">
<li>
    <a class="dropdown-item" href="../../syllabus.html">
 <span class="dropdown-text">Syllabus</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../schedule.html">
 <span class="dropdown-text">Schedule</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-course-materials" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Course Materials</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-course-materials">
<li>
    <a class="dropdown-item" href="../../lectures.html">
 <span class="dropdown-text">Lectures</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../projects.html">
 <span class="dropdown-text">Projects</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../resources.html">
 <span class="dropdown-text">Resources</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/mbutler808"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/mbutler808"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">The ggtree-verse</h1>
                  <div>
        <div class="description">
          Working with trees with tidytree and plotting with ggtree
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">module 7</div>
                <div class="quarto-category">week 13</div>
                <div class="quarto-category">phylogenetic trees</div>
                <div class="quarto-category">phylo</div>
                <div class="quarto-category">ggtree</div>
                <div class="quarto-category">treedata</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliation</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author"><a href="https://butlerlab.org">Marguerite Butler</a> </p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              School of Life Sciences, University of Hawaii
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">April 23, 2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link active" data-scroll-target="#acknowledgements">Acknowledgements</a></li>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning objectives</a></li>
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview">Overview</a></li>
  <li>
<a href="#phylo---treedata" id="toc-phylo---treedata" class="nav-link" data-scroll-target="#phylo---treedata">phylo -&gt; treedata</a>
  <ul class="collapse">
<li><a href="#coverting-between-data-types" id="toc-coverting-between-data-types" class="nav-link" data-scroll-target="#coverting-between-data-types">Coverting between data types</a></li>
  </ul>
</li>
  <li>
<a href="#simulate-phenotypic-data" id="toc-simulate-phenotypic-data" class="nav-link" data-scroll-target="#simulate-phenotypic-data">Simulate phenotypic data</a>
  <ul class="collapse">
<li><a href="#add-tiplabels" id="toc-add-tiplabels" class="nav-link" data-scroll-target="#add-tiplabels">Add tiplabels</a></li>
  <li><a href="#get-the-tip-labels" id="toc-get-the-tip-labels" class="nav-link" data-scroll-target="#get-the-tip-labels">Get the tip labels</a></li>
  </ul>
</li>
  <li>
<a href="#merge-tree-with-data" id="toc-merge-tree-with-data" class="nav-link" data-scroll-target="#merge-tree-with-data">Merge tree with data</a>
  <ul class="collapse">
<li><a href="#subsetting-the-tree" id="toc-subsetting-the-tree" class="nav-link" data-scroll-target="#subsetting-the-tree">Subsetting the tree</a></li>
  <li><a href="#plotting-with-node-labels" id="toc-plotting-with-node-labels" class="nav-link" data-scroll-target="#plotting-with-node-labels">Plotting with node labels</a></li>
  <li><a href="#plotting-with-alternative-tip-labels" id="toc-plotting-with-alternative-tip-labels" class="nav-link" data-scroll-target="#plotting-with-alternative-tip-labels">Plotting with alternative tip labels</a></li>
  <li><a href="#when-your-tip-labels-get-cut-off" id="toc-when-your-tip-labels-get-cut-off" class="nav-link" data-scroll-target="#when-your-tip-labels-get-cut-off">When your tip labels get cut off</a></li>
  </ul>
</li>
  <li><a href="#tree-layouts" id="toc-tree-layouts" class="nav-link" data-scroll-target="#tree-layouts">Tree layouts</a></li>
  <li>
<a href="#plotting-data-on-the-tree" id="toc-plotting-data-on-the-tree" class="nav-link" data-scroll-target="#plotting-data-on-the-tree">Plotting data on the tree</a>
  <ul class="collapse">
<li><a href="#example-plot-smalltree-with-size-in-a-barplot" id="toc-example-plot-smalltree-with-size-in-a-barplot" class="nav-link" data-scroll-target="#example-plot-smalltree-with-size-in-a-barplot">Example: plot <code>smalltree</code> with <code>size</code> in a barplot</a></li>
  </ul>
</li>
  <li>
<a href="#a-tree-panel-and-annotation-example-from-the-tree-data-book" id="toc-a-tree-panel-and-annotation-example-from-the-tree-data-book" class="nav-link" data-scroll-target="#a-tree-panel-and-annotation-example-from-the-tree-data-book">A tree-panel and annotation example from the Tree Data Book:</a>
  <ul class="collapse">
<li><a href="#the-operator-for-ggtree-objects" id="toc-the-operator-for-ggtree-objects" class="nav-link" data-scroll-target="#the-operator-for-ggtree-objects">The %+&gt;% operator for ggtree objects</a></li>
  <li>
<a href="#example-of-the-operator-to-add-data-to-a-ggtree-object." id="toc-example-of-the-operator-to-add-data-to-a-ggtree-object." class="nav-link" data-scroll-target="#example-of-the-operator-to-add-data-to-a-ggtree-object.">Example of the %+&gt;% operator to add data to a ggtree object.</a>
  <ul class="collapse">
<li><a href="#explore-df_info" id="toc-explore-df_info" class="nav-link" data-scroll-target="#explore-df_info">Explore <code>df_info</code></a></li>
  <li><a href="#add-snp-and-trait-plots-aligned-to-the-tree" id="toc-add-snp-and-trait-plots-aligned-to-the-tree" class="nav-link" data-scroll-target="#add-snp-and-trait-plots-aligned-to-the-tree">Add SNP and Trait plots aligned to the tree</a></li>
  </ul>
</li>
  </ul>
</li>
  <li><a href="#example-datasets" id="toc-example-datasets" class="nav-link" data-scroll-target="#example-datasets">Example datasets</a></li>
  </ul></nav>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><section id="acknowledgements" class="level3"><h3 class="anchored" data-anchor-id="acknowledgements">Acknowledgements</h3>
<p>References for this Material:</p>
<ul>
<li>Treedata book by Guangchuang Yu <span class="citation" data-cites="Yu:2022">(<a href="#ref-Yu:2022" role="doc-biblioref">Yu 2022</a>)</span> <a href="https://yulab-smu.top/treedata-book/" class="uri">https://yulab-smu.top/treedata-book/</a>
</li>
<li><a href="https://bioconnector.github.io/workshops/r-ggtree.html" class="uri">https://bioconnector.github.io/workshops/r-ggtree.html</a></li>
</ul></section><section id="learning-objectives" class="level1"><h1>Learning objectives</h1>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>At the end of this lesson you will:</strong></p>
<ul>
<li>Understand information content of phylogenetically structured data</li>
<li>Understand particular R tree formats in ape, phylobase, and ouch</li>
<li>Be able to hand-make trees</li>
<li>Be able to import trees from nexus, newick, and other major formats in use today</li>
<li>Be able to convert trees from one format to another</li>
<li>Be able to perform basic tree manipulations</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
R packages you will need
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><code>ape</code></li>
<li><code>ggtree</code></li>
<li><code>tidytree</code></li>
<li><code>treeio</code></li>
<li><code>TDbook</code></li>
<li><code>ggplot2</code></li>
<li><code>ouch</code></li>
</ul>
</div>
</div>
</section><section id="overview" class="level1"><h1>Overview</h1>
<p><code>ggtree</code> is a powerful phylogenetic tree plotting package, that is used alongside <code>treeio</code> and <code>tidytree</code> packages to import/export and to manipulate phylogenetic trees, respectively. These packages allow you to subset or combine trees with data, annotate, and plot in so many different ways using the grammar of graphics.</p>
<p>Because itÊ»s written in the style of <code>ggplot2</code> and the <code>tidyverse</code>, there are a lot of functions to remember, and now there are also multiple object types. The key to working effectively with <code>ggtree</code> is to be aware of what type of object you are working with, whether it is a <code>dataframe</code>, a <code>tibble</code>, a <code>phylo</code> tree, a <code>treedata</code> tibble, or <code>ggtree</code> object. It is easy to convert between these at will if you know what you are dealing with.</p>
</section><section id="phylo---treedata" class="level1"><h1>phylo -&gt; treedata</h1>
<p>Functions like <code>read.tree</code> amd <code>read.nexus</code>, etc. will read in objects of class <code>phylo</code> (they are actually referencing the ape function).</p>
<p>To show this, letÊ»s first generate a random tree using <code><a href="https://rdrr.io/pkg/ape/man/rtree.html">ape::rtree()</a></code></p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/">ggtree</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/contribution-tree-data/">treeio</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/rtree.html">rtree</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phylo"</code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>  <span class="co"># ape plotting function</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Note: <code>ggtree</code> can also accept <code>phylo</code> objects as arguments:</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>  <span class="co"># ggtree plotting function</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>To save as newick and nexus formats, <code>treeio</code> has the following:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">treeio</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.nexus.html">write.nexus</a></span><span class="op">(</span><span class="va">tree</span>, file<span class="op">=</span><span class="st">"tree.nex"</span><span class="op">)</span></span>
<span><span class="fu">treeio</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html">write.tree</a></span><span class="op">(</span><span class="va">tree</span>, file<span class="op">=</span><span class="st">"tree.tree"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "_index.qmd"         "anolis.SSD.raw.csv" "bigtree.nex"       
 [4] "ggtree_functions.R" "ggtree.R"           "inclass.R"         
 [7] "index_files"        "index.qmd"          "index.rmarkdown"   
[10] "tree.nex"           "tree.tree"         </code></pre>
</div>
</div>
<p>Take a look at these files. They should be in nexus and newick formats, respectively.</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tree_nex</span> <span class="op">&lt;-</span> <span class="fu">treeio</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.nexus.html">read.nexus</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"tree.nex"</span><span class="op">)</span></span>
<span><span class="va">tree_new</span> <span class="op">&lt;-</span> <span class="fu">treeio</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/treeio/man/read.newick.html">read.newick</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"tree.tree"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">tree_nex</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phylo"</code></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">tree_new</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phylo"</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you have trees from <code>iqtree</code>, <code>BEAST</code> etc., you should use the specialized read functions because they will capture the metadata embedded in the treeformat including support values.</p>
</div>
</div>
<section id="coverting-between-data-types" class="level2"><h2 class="anchored" data-anchor-id="coverting-between-data-types">Coverting between data types</h2>
<p>The main functions for coersion are:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.phylo.html">as.phylo</a></span> <span class="op">(</span><span class="va">tree</span><span class="op">)</span>  <span class="co"># to phylo</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 20 tips and 19 internal nodes.

Tip labels:
  t6, t19, t15, t16, t2, t18, ...

Rooted; includes branch length(s).</code></pre>
</div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/tidytree/man/as.treedata.html">as.treedata</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="co"># to treedata</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 20 tips and 19 internal nodes.

Tip labels:
  t6, t19, t15, t16, t2, t18, ...

Rooted; includes branch length(s).</code></pre>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">treedata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tidytree/man/as.treedata.html">as.treedata</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>  </span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">treedata</span><span class="op">)</span>  <span class="co"># to tibble</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tbl_tree abstraction: 39 Ã— 4
# which can be converted to treedata or phylo 
# via as.treedata or as.phylo
   parent  node branch.length label
    &lt;int&gt; &lt;int&gt;         &lt;dbl&gt; &lt;chr&gt;
 1     21     1        0.211  t6   
 2     25     2        0.0205 t19  
 3     25     3        0.493  t15  
 4     26     4        0.359  t16  
 5     26     5        0.363  t2   
 6     27     6        0.616  t18  
 7     27     7        0.287  t13  
 8     31     8        0.540  t14  
 9     32     9        0.202  t10  
10     34    10        0.0719 t9   
# â„¹ 29 more rows</code></pre>
</div>
</div>
</section></section><section id="simulate-phenotypic-data" class="level1"><h1>Simulate phenotypic data</h1>
<section id="add-tiplabels" class="level3"><h3 class="anchored" data-anchor-id="add-tiplabels">Add tiplabels</h3>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="get-the-tip-labels" class="level3"><h3 class="anchored" data-anchor-id="get-the-tip-labels">Get the tip labels</h3>
<p>There is a function <code><a href="https://rdrr.io/pkg/ggtree/man/get_taxa_name.html">get_taxa_name()</a></code> which is very handy, but NOTE: its argument is a <code>ggtree</code> object, not a <code>treedata</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/get_taxa_name.html">get_taxa_name</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "t17" "t5"  "t12" "t1"  "t9"  "t10" "t14" "t11" "t20" "t8"  "t3"  "t4" 
[13] "t7"  "t2"  "t16" "t15" "t19" "t13" "t18" "t6" </code></pre>
</div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/get_taxa_name.html">get_taxa_name</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Simulate a data matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="va">size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">n</span>, mean<span class="op">=</span><span class="fl">20</span>, sd<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">habitat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"desert"</span>, <span class="st">"grassland"</span>, <span class="st">"forest"</span>, <span class="st">"intertidal"</span><span class="op">)</span>, size<span class="op">=</span><span class="va">n</span>, replace<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span> <span class="st">"label"</span><span class="op">=</span> <span class="va">taxa</span>, <span class="va">size</span>, <span class="va">habitat</span><span class="op">)</span></span>
<span><span class="va">dat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   label     size    habitat
1    t17 12.86105  grassland
2     t5 31.25774     forest
3    t12 21.43777  grassland
4     t1 17.61510  grassland
5     t9 21.89762     forest
6    t10 18.52082     desert
7    t14 30.01839  grassland
8    t11 23.58696  grassland
9    t20 23.90757 intertidal
10    t8 24.53455     desert
11    t3 21.79389     desert
12    t4 19.42923  grassland
13    t7 21.90441     desert
14    t2 27.65435     desert
15   t16 14.96855 intertidal
16   t15 24.14979     forest
17   t19 12.13357  grassland
18   t13 28.80982     forest
19   t18 19.61751     desert
20    t6 17.84677 intertidal</code></pre>
</div>
</div>
<p>See our ggtree as a treedata object:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/tidytree/man/as.treedata.html">as.treedata</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">as_tibble</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">as.data.frame</span>  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   parent node branch.length label
1      21    1   0.211462186    t6
2      25    2   0.020523685   t19
3      25    3   0.492876891   t15
4      26    4   0.358771561   t16
5      26    5   0.362600771    t2
6      27    6   0.616498912   t18
7      27    7   0.286937086   t13
8      31    8   0.539667591   t14
9      32    9   0.201774493   t10
10     34   10   0.071909918    t9
11     34   11   0.548395909    t1
12     35   12   0.781278409   t12
13     36   13   0.834200829    t5
14     36   14   0.865846023   t17
15     30   15   0.591518627   t11
16     29   16   0.317440855   t20
17     37   17   0.241390239    t7
18     38   18   0.517082721    t4
19     39   19   0.385656054    t3
20     39   20   0.519972739    t8
21     21   21            NA  &lt;NA&gt;
22     21   22   0.104038035  &lt;NA&gt;
23     22   23   0.160041574  &lt;NA&gt;
24     23   24   0.095194947  &lt;NA&gt;
25     24   25   0.010641325  &lt;NA&gt;
26     24   26   0.358477827  &lt;NA&gt;
27     23   27   0.563482889  &lt;NA&gt;
28     22   28   0.009179101  &lt;NA&gt;
29     28   29   0.870633200  &lt;NA&gt;
30     29   30   0.373868619  &lt;NA&gt;
31     30   31   0.181675595  &lt;NA&gt;
32     31   32   0.950593904  &lt;NA&gt;
33     32   33   0.999016533  &lt;NA&gt;
34     33   34   0.925645936  &lt;NA&gt;
35     33   35   0.341042848  &lt;NA&gt;
36     35   36   0.493788643  &lt;NA&gt;
37     28   37   0.938912940  &lt;NA&gt;
38     37   38   0.628192976  &lt;NA&gt;
39     38   39   0.066866069  &lt;NA&gt;</code></pre>
</div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span>  <span class="co"># ggtree -&gt; treedata -&gt; tibble -&gt; dataframe</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="merge-tree-with-data" class="level1"><h1>Merge tree with data</h1>
<p>Now that we have a matching key in both the tree and data objects, we can join the tree with the dataframe by those matching labels using <code>ggtree</code>Ê»s <code>full_join</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ttree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">full_join</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">dat</span>, by <span class="op">=</span> <span class="st">"label"</span><span class="op">)</span> </span>
<span><span class="va">ttree</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 20 tips and 19 internal nodes.

Tip labels:
  t6, t19, t15, t16, t2, t18, ...

Rooted; includes branch length(s).

with the following features available:
  '', 'size', 'habitat'.

# The associated data tibble abstraction: 39 Ã— 5
# The 'node', 'label' and 'isTip' are from the phylo tree.
    node label isTip  size habitat   
   &lt;int&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;chr&gt;     
 1     1 t6    TRUE   17.8 intertidal
 2     2 t19   TRUE   12.1 grassland 
 3     3 t15   TRUE   24.1 forest    
 4     4 t16   TRUE   15.0 intertidal
 5     5 t2    TRUE   27.7 desert    
 6     6 t18   TRUE   19.6 desert    
 7     7 t13   TRUE   28.8 forest    
 8     8 t14   TRUE   30.0 grassland 
 9     9 t10   TRUE   18.5 desert    
10    10 t9    TRUE   21.9 forest    
# â„¹ 29 more rows</code></pre>
</div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ttree</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">as_tibble</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">as.data.frame</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   parent node branch.length label     size    habitat
1      21    1   0.211462186    t6 17.84677 intertidal
2      25    2   0.020523685   t19 12.13357  grassland
3      25    3   0.492876891   t15 24.14979     forest
4      26    4   0.358771561   t16 14.96855 intertidal
5      26    5   0.362600771    t2 27.65435     desert
6      27    6   0.616498912   t18 19.61751     desert
7      27    7   0.286937086   t13 28.80982     forest
8      31    8   0.539667591   t14 30.01839  grassland
9      32    9   0.201774493   t10 18.52082     desert
10     34   10   0.071909918    t9 21.89762     forest
11     34   11   0.548395909    t1 17.61510  grassland
12     35   12   0.781278409   t12 21.43777  grassland
13     36   13   0.834200829    t5 31.25774     forest
14     36   14   0.865846023   t17 12.86105  grassland
15     30   15   0.591518627   t11 23.58696  grassland
16     29   16   0.317440855   t20 23.90757 intertidal
17     37   17   0.241390239    t7 21.90441     desert
18     38   18   0.517082721    t4 19.42923  grassland
19     39   19   0.385656054    t3 21.79389     desert
20     39   20   0.519972739    t8 24.53455     desert
21     21   21            NA  &lt;NA&gt;       NA       &lt;NA&gt;
22     21   22   0.104038035  &lt;NA&gt;       NA       &lt;NA&gt;
23     22   23   0.160041574  &lt;NA&gt;       NA       &lt;NA&gt;
24     23   24   0.095194947  &lt;NA&gt;       NA       &lt;NA&gt;
25     24   25   0.010641325  &lt;NA&gt;       NA       &lt;NA&gt;
26     24   26   0.358477827  &lt;NA&gt;       NA       &lt;NA&gt;
27     23   27   0.563482889  &lt;NA&gt;       NA       &lt;NA&gt;
28     22   28   0.009179101  &lt;NA&gt;       NA       &lt;NA&gt;
29     28   29   0.870633200  &lt;NA&gt;       NA       &lt;NA&gt;
30     29   30   0.373868619  &lt;NA&gt;       NA       &lt;NA&gt;
31     30   31   0.181675595  &lt;NA&gt;       NA       &lt;NA&gt;
32     31   32   0.950593904  &lt;NA&gt;       NA       &lt;NA&gt;
33     32   33   0.999016533  &lt;NA&gt;       NA       &lt;NA&gt;
34     33   34   0.925645936  &lt;NA&gt;       NA       &lt;NA&gt;
35     33   35   0.341042848  &lt;NA&gt;       NA       &lt;NA&gt;
36     35   36   0.493788643  &lt;NA&gt;       NA       &lt;NA&gt;
37     28   37   0.938912940  &lt;NA&gt;       NA       &lt;NA&gt;
38     37   38   0.628192976  &lt;NA&gt;       NA       &lt;NA&gt;
39     38   39   0.066866069  &lt;NA&gt;       NA       &lt;NA&gt;</code></pre>
</div>
</div>
<p>And thatÊ»s what our <code>treedata</code> object looks like flattened out!</p>
<section id="subsetting-the-tree" class="level2"><h2 class="anchored" data-anchor-id="subsetting-the-tree">Subsetting the tree</h2>
<p>Functions: <code><a href="https://rdrr.io/pkg/ape/man/drop.tip.html">drop.tip()</a></code> and <code><a href="https://rdrr.io/pkg/ape/man/drop.tip.html">keep.tip()</a></code></p>
<p>Suppose we want to drop all of the even tips:</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">todrop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"t"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">*</span><span class="fl">2</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span><span class="va">todrop</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "t2"  "t4"  "t6"  "t8"  "t10" "t12" "t14" "t16" "t18" "t20"</code></pre>
</div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">smalltree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/drop.tip.html">drop.tip</a></span><span class="op">(</span><span class="va">ttree</span>, <span class="va">todrop</span><span class="op">)</span></span>
<span><span class="va">smalltree</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'treedata' S4 object'.

...@ phylo:

Phylogenetic tree with 10 tips and 9 internal nodes.

Tip labels:
  t19, t15, t13, t9, t1, t5, ...

Rooted; includes branch length(s).

with the following features available:
  '', 'size', 'habitat'.

# The associated data tibble abstraction: 19 Ã— 5
# The 'node', 'label' and 'isTip' are from the phylo tree.
    node label isTip  size habitat  
   &lt;int&gt; &lt;chr&gt; &lt;lgl&gt; &lt;dbl&gt; &lt;chr&gt;    
 1     1 t19   TRUE   12.1 grassland
 2     2 t15   TRUE   24.1 forest   
 3     3 t13   TRUE   28.8 forest   
 4     4 t9    TRUE   21.9 forest   
 5     5 t1    TRUE   17.6 grassland
 6     6 t5    TRUE   31.3 forest   
 7     7 t17   TRUE   12.9 grassland
 8     8 t11   TRUE   23.6 grassland
 9     9 t7    TRUE   21.9 desert   
10    10 t3    TRUE   21.8 desert   
# â„¹ 9 more rows</code></pre>
</div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">smalltree</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><code>drop.tip</code> keeps all of the metadata! <code>keep.tip</code> is imported from <code>ape</code> so it has to be converted to <code>phylo</code> and then the data joined again after.</p>
</section><section id="plotting-with-node-labels" class="level2"><h2 class="anchored" data-anchor-id="plotting-with-node-labels">Plotting with node labels</h2>
<p>The geometries <code><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text()</a></code> and <code>geom_node()</code> are helpful for labelling all of the nodes. The function <code><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab()</a></code> labels only the tips.</p>
<p>Add node labels so you know what the internal node numbers are:</p>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">smalltree</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">node</span><span class="op">)</span>, hjust<span class="op">=</span><span class="op">-</span><span class="fl">.3</span><span class="op">)</span>    <span class="co"># node numbers</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Note: The tiplabels and the node labels crashed!</p>
<p>There are also <code>2</code> versions: <code><a href="https://rdrr.io/pkg/ggtree/man/geom_text2.html">geom_text2()</a></code> and <code>geom_node2()</code> that allow subsetting the nodes, when you want the geometry to apply to only some of the nodes.</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">smalltree</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_text2.html">geom_text2</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">node</span>, subset<span class="op">=</span><span class="op">!</span><span class="va">isTip</span><span class="op">)</span>, hjust<span class="op">=</span><span class="op">-</span><span class="fl">.3</span><span class="op">)</span>    <span class="co"># node numbers</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><code>isTip</code> is a column of the ggtree object, so it is inherited when we provide the ggtree object.</p>
</section><section id="plotting-with-alternative-tip-labels" class="level2"><h2 class="anchored" data-anchor-id="plotting-with-alternative-tip-labels">Plotting with alternative tip labels</h2>
<p>The dataframe portion of the treedata object can hold any number of columns of metadata. Perhaps you have some real names in a different column (like a display name), it is easy to swap out the tip labels. Here letÊ»s just use the <code>habitat</code> column</p>
<div class="cell">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">smalltree</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">habitat</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_text2.html">geom_text2</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">node</span>, subset<span class="op">=</span><span class="op">!</span><span class="va">isTip</span><span class="op">)</span>, hjust<span class="op">=</span><span class="op">-</span><span class="fl">.3</span><span class="op">)</span>    <span class="co"># node numbers</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="when-your-tip-labels-get-cut-off" class="level2"><h2 class="anchored" data-anchor-id="when-your-tip-labels-get-cut-off">When your tip labels get cut off</h2>
<p>Add an x scale (usually time):</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">smalltree</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">habitat</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_text2.html">geom_text2</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">node</span>, subset<span class="op">=</span><span class="op">!</span><span class="va">isTip</span><span class="op">)</span>, hjust<span class="op">=</span><span class="op">-</span><span class="fl">.3</span><span class="op">)</span> <span class="op">+</span>   <span class="co"># node numbers</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/theme_tree2.html">theme_tree2</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You can increase the size of the plot area to accommodate the longer labels:</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">smalltree</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">habitat</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_text2.html">geom_text2</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">node</span>, subset<span class="op">=</span><span class="op">!</span><span class="va">isTip</span><span class="op">)</span>, hjust<span class="op">=</span><span class="op">-</span><span class="fl">.3</span><span class="op">)</span> <span class="op">+</span>   <span class="co"># node numbers</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/theme_tree2.html">theme_tree2</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="tree-layouts" class="level1"><h1>Tree layouts</h1>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span>, branch.length<span class="op">=</span><span class="st">'none'</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span>, layout<span class="op">=</span><span class="st">"dendrogram"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span>, layout<span class="op">=</span><span class="st">"roundrect"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span>, layout<span class="op">=</span><span class="st">"ellipse"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span>, layout<span class="op">=</span><span class="st">"ellipse"</span>, branch.length<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span>, layout<span class="op">=</span><span class="st">"circular"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span>, branch.length<span class="op">=</span><span class="st">'none'</span>, layout<span class="op">=</span><span class="st">'circular'</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span>, layout<span class="op">=</span><span class="st">"fan"</span>, open.angle<span class="op">=</span><span class="fl">120</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">ttree</span>, layout<span class="op">=</span><span class="st">"inward_circular"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="plotting-data-on-the-tree" class="level1"><h1>Plotting data on the tree</h1>
<p><code><a href="https://rdrr.io/pkg/ggtree/man/facet-plot.html">geom_facet()</a></code> and <code><a href="https://rdrr.io/pkg/ggtree/man/facet-plot.html">facet_plot()</a></code> are general methods to link graphical layers to a tree.</p>
<p>These functions require an input dataframe with the first column containing the taxon labels (the key which matches to the tip labels of the phylogeny).</p>
<p>Internally these functions reorder the input data based on the tree structure so that you donÊ»t have to worry about the order of the rows.</p>
<p>Multiple layers can be added to the same dataset. Also different datasets can be added to the same figure.</p>
<p>A table of the <strong>geom layers</strong> that work with <code>geom_facet</code> is <a href="https://yulab-smu.top/treedata-book/figures-and-tables.html#tab:facet-geom">provided here</a>.</p>
<section id="example-plot-smalltree-with-size-in-a-barplot" class="level2"><h2 class="anchored" data-anchor-id="example-plot-smalltree-with-size-in-a-barplot">Example: plot <code>smalltree</code> with <code>size</code> in a barplot</h2>
<p>First make a tibble to attach to the tree. As of this writing, <code>geom_facet</code> will not accept a <code>treedata</code> object. It wants a dataframe or tibble of only the tips. But this is easy to make from the treedata. We just have to filter out the non-tip rows, then rearrange the columns to put the labels first:</p>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">smdat</span> <span class="op">&lt;-</span> <span class="va">smalltree</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>           <span class="va">as_tibble</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">label</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html">relocate</a></span><span class="op">(</span><span class="va">label</span>, .before<span class="op">=</span><span class="fl">1</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then add the barplot as a panel next to the tree plot using <code>geom_facet</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">smalltree</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/theme_tree2.html">theme_tree2</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/facet-plot.html">geom_facet</a></span><span class="op">(</span>panel <span class="op">=</span> <span class="st">"Size"</span>, </span>
<span>               data<span class="op">=</span><span class="va">smdat</span>, </span>
<span>               geom <span class="op">=</span> <span class="va">geom_col</span>, </span>
<span>               mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">smdat</span><span class="op">$</span><span class="va">size</span><span class="op">)</span>, </span>
<span>               orientation <span class="op">=</span> <span class="st">'y'</span>, </span>
<span>               width <span class="op">=</span> <span class="fl">.6</span>, </span>
<span>               fill<span class="op">=</span><span class="st">"blue"</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The arguments for geom_facet() are:</p>
<ul>
<li>
<code>panel</code> : The name of the panel, displayed on top</li>
<li>
<code>data</code> : a <code>tibble</code> or <code>dataframe</code> containing the metadata to plot. Must have as the first column the tip labels that are found in the phylogenetic tree.</li>
<li>
<code>geom</code> : a <a href="https://yulab-smu.top/treedata-book/figures-and-tables.html#tab:facet-geom">geom layer</a> specifying the style of plot</li>
<li>
<code>mapping</code> : the aesthetic mapping. I should not have to supply the <code>smdat$</code> here but it wonÊ»t work otherwise.</li>
<li>any additional parameters for the plot</li>
</ul></section></section><section id="a-tree-panel-and-annotation-example-from-the-tree-data-book" class="level1"><h1>A tree-panel and annotation example from the Tree Data Book:</h1>
<p>This example plots a phylogeny alongside SNP (single nucleotide polymorphism) data and a barplot of some simulated data <span class="citation" data-cites="Yu:2022">(<a href="#ref-Yu:2022" role="doc-biblioref">Yu 2022</a>)</span>.</p>
<section id="the-operator-for-ggtree-objects" class="level2"><h2 class="anchored" data-anchor-id="the-operator-for-ggtree-objects">The %+&gt;% operator for ggtree objects</h2>
<p>The %+&gt;% operator is used to add data (<code>dataframe</code>, <code>tibble</code>) to a <code>ggtree</code> object:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="ex">my_ggtree</span> <span class="op">&lt;</span>- my_ggtree %<span class="op">&lt;</span>+% new_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The result is a combined object that can be used for plotting, but it does not modify the original <code>treedata</code> object (which is a different object from the <code>ggtree</code> object). The <code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">full_join()</a></code> function can be used to combine a tree with data to produce a new <code>treedata</code> object.</p>
</section><section id="example-of-the-operator-to-add-data-to-a-ggtree-object." class="level2"><h2 class="anchored" data-anchor-id="example-of-the-operator-to-add-data-to-a-ggtree-object.">Example of the %+&gt;% operator to add data to a ggtree object.</h2>
<p>The package <code>TDbook</code> is the data accompanyment to <span class="citation" data-cites="Yu:2022">(<a href="#ref-Yu:2022" role="doc-biblioref">Yu 2022</a>)</span>Ê»s <a href="https://yulab-smu.top/treedata-book/index.html">Tree Data book</a>. It is available on CRAN so you can install it with the usual <code>install.packages("TDbook")</code> function call.</p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/">ggtree</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/">TDbook</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load `tree_boots`, `df_tip_data`, and `df_inode_data` from 'TDbook'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree_boots</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/ggfun/man/attacher.html">%&lt;+%</a></span> <span class="va">df_tip_data</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">.1</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span>offset <span class="op">=</span> <span class="fl">.6</span>, hjust <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">trophic_habit</span>, color <span class="op">=</span> <span class="va">trophic_habit</span>, </span>
<span>                size <span class="op">=</span> <span class="va">mass_in_kg</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op"><a href="https://rdrr.io/pkg/ggfun/man/attacher.html">%&lt;+%</a></span> <span class="va">df_inode_data</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_label</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">vernacularName.y</span>, fill <span class="op">=</span> <span class="va">posterior</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"YlGnBu"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 7 rows containing missing values or values outside the scale range
(`geom_label()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="explore-df_info" class="level3"><h3 class="anchored" data-anchor-id="explore-df_info">Explore <code>df_info</code>
</h3>
<p><code>df_info</code> A dataframe containing sampling info for the tips of the tree. 386 rows and 6 variables, with the first column being taxa labels (<code>id</code>).</p>
<p><code>df_alleles</code> The allele table with original raw data to be processed to SNP data. It is a table of nucleotides with 386 rows x 385 variables. The first row contains tips labels. Column names are non-sense. The rownames (exept for the first one) contains the snp position along the genome.</p>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## load `tree_nwk`, `df_info`, `df_alleles`, and `df_bar_data` from 'TDbook'</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="va">tree_nwk</span></span>
<span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="va">df_alleles</span></span>
<span><span class="va">snps_strainCols</span> <span class="op">&lt;-</span> <span class="va">snps</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span> </span>
<span><span class="va">snps</span><span class="op">&lt;-</span><span class="va">snps</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">]</span> <span class="co"># drop strain names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">snps_strainCols</span></span>
<span></span>
<span><span class="va">gapChar</span> <span class="op">&lt;-</span> <span class="st">"?"</span></span>
<span><span class="va">snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">snps</span><span class="op">)</span>  <span class="co"># not rectangular!</span></span>
<span><span class="va">lsnp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">snp</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">x</span> <span class="op">!=</span> <span class="va">snp</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span> <span class="op">&amp;</span> <span class="va">x</span> <span class="op">!=</span> <span class="va">gapChar</span> <span class="op">&amp;</span> <span class="va">snp</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span> <span class="op">!=</span> <span class="va">gapChar</span></span>
<span>    <span class="op">}</span><span class="op">)</span>  <span class="co"># different from row 1, not missing</span></span>
<span><span class="va">lsnp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">lsnp</span><span class="op">)</span> </span>
<span><span class="va">lsnp</span><span class="op">$</span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">lsnp</span><span class="op">)</span><span class="op">)</span>  <span class="co"># position from rownames</span></span>
<span><span class="va">lsnp</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span><span class="op">(</span><span class="va">lsnp</span>, <span class="va">name</span>, <span class="va">value</span>, <span class="op">-</span><span class="va">pos</span><span class="op">)</span>  </span>
<span><span class="va">snp_data</span> <span class="op">&lt;-</span> <span class="va">lsnp</span><span class="op">[</span><span class="va">lsnp</span><span class="op">$</span><span class="va">value</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"name"</span>, <span class="st">"pos"</span><span class="op">)</span><span class="op">]</span> <span class="co"># only TRUEs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>snp_data</code> A dataframe containing SNP position data. 6482 x 2. The first column contains taxa labels coresponding to the tips of the tree (<code>name</code>). There are multiple rows per taxon, the second colum is the position <code>pos</code> of the snp in the genome. This is used as the x-variable in the plot.</p>
<p>In the dataframe <code>snp_data</code> the rows are ordered by position along the sequence (the x-dimension of this data), but the first column is the strain (taxon) name which matches the tips in the phylogenetic tree.</p>
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## visualize the tree </span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">## attach the sampling information data set </span></span>
<span><span class="co">## and add symbols colored by location</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op"><a href="https://rdrr.io/pkg/ggfun/man/attacher.html">%&lt;+%</a></span> <span class="va">df_info</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">location</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="add-snp-and-trait-plots-aligned-to-the-tree" class="level3"><h3 class="anchored" data-anchor-id="add-snp-and-trait-plots-aligned-to-the-tree">Add SNP and Trait plots aligned to the tree</h3>
<p>Use <code>geom_facet</code> with reference to the respective dataframes/tibbles to add plots alignted to the tree. For the SNP plot, we will use <code>geom_point</code> which allows x-y plotting, with x-coordinate determined by <code>pos</code> and the y-coordinate aligned with the taxon. The symbol is the vertical line <code>|</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## visualize SNP and Trait data using dot and bar charts,</span></span>
<span><span class="co">## and align them based on tree structure</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/facet-plot.html">geom_facet</a></span><span class="op">(</span>panel <span class="op">=</span> <span class="st">"SNP"</span>, data <span class="op">=</span> <span class="va">snp_data</span>, geom <span class="op">=</span> <span class="va">geom_point</span>, </span>
<span>               mapping<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pos</span>, color <span class="op">=</span> <span class="va">location</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">'|'</span><span class="op">)</span> </span>
<span><span class="va">p1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><code>df_bar_data</code> is some simulated data with an <code>id</code> column specifying the taxon names, and a <code>dummy_bar_value</code> containing some data.</p>
<div class="cell">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/facet-plot.html">geom_facet</a></span><span class="op">(</span>panel <span class="op">=</span> <span class="st">"Trait"</span>, data <span class="op">=</span> <span class="va">df_bar_data</span>, geom <span class="op">=</span> <span class="va">geom_col</span>, </span>
<span>                <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dummy_bar_value</span>, color <span class="op">=</span> <span class="va">location</span>, </span>
<span>                fill <span class="op">=</span> <span class="va">location</span><span class="op">)</span>, orientation <span class="op">=</span> <span class="st">'y'</span>, width <span class="op">=</span> <span class="fl">.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/theme_tree2.html">theme_tree2</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.05</span>, <span class="fl">.85</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section></section><section id="example-datasets" class="level1"><h1>Example datasets</h1>
<p>save to your working directory:</p>
<p><a href="bigtree.nex">bigtree.nex</a><br><a href="anolis.SSD.raw.csv">anolis.SSD.raw.csv</a><br><a href="ggtree.R">ggtree.R</a></p>
<p>This is an example of a typical workflow. We have carefully collected phenotypic data, and someone has published a massive phylogeny. We need to subset the tree to just the taxa we want to work on.</p>



</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Yu:2022" class="csl-entry" role="listitem">
Yu, G. 2022. <em>Data Integration, Manipulation and Visualization of Phylogenetic Trees</em>. A Chapman &amp; Hall Book. CRC Press. <a href="https://yulab-smu.top/treedata-book/">https://yulab-smu.top/treedata-book/</a>.
</div>
</div></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/mbutler808\.github\.io");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>